google/cloud/bigtable/async_read_stream_test.cc

```diff
@@ -0,0 +1,723 @@
+  auto op = cq_.MakeStreamingReadRpc(
+      [this](grpc::ClientContext* context,
+             btproto::MutateRowsRequest const& request,
+             grpc::CompletionQueue* cq) {
+        return stub_->PrepareAsyncMutateRows(context, request, cq);
+      },
+      request, std::move(context), on_read,
+      [&result](Status s) {
+        result.status = std::move(s);
+        result.done.Lift();
+      });
+
+  server_sent_responses_barrier.Wait();
+  read_received_barrier.Wait();
+  op->Cancel();
+  op->Cancel();
+  cancel_done_server_barrier.Lift();
+  cancel_done_read_barrier.Lift();
+
+  // The server remains blocked until the stream finishes, therefore, the only
+  // way this actually unblocks is if the Cancel() succeeds.
+  result.done.Wait();
+  ASSERT_EQ(2U, result.reads.size());
+  EXPECT_EQ(StatusCode::kCancelled, result.status.code());
+  for (int i = 0; i != 2; ++i) {
+    SCOPED_TRACE("Running iteration: " + std::to_string(i));
+    ASSERT_EQ(1, result.reads[0].entries_size());
+    EXPECT_EQ(0, result.reads[0].entries(0).index());
+  }
+
+  // The barriers go out of scope when this function exits, but the server may
+  // still be using them, so wait for the server to shutdown before leaving the
+  // scope.
+  WaitForServerShutdown();
+}
+
+/// @test Verify that AsyncReadStream works when one Cancels() before reading.
+TEST_F(AsyncReadStreamTest, CancelBeforeRead) {
+  SimpleBarrier server_started_barrier;
+  SimpleBarrier cancel_done_server_barrier;
+  impl_.SetCallback(
+      [this, &server_started_barrier, &cancel_done_server_barrier](
+          grpc::ServerContext*, google::bigtable::v2::MutateRowsRequest const*,
+          grpc::ServerWriter<google::bigtable::v2::MutateRowsResponse>*
+              writer) {
+        server_started_barrier.Lift();
+        WriteOne(writer, 0);
+        WriteOne(writer, 1);
+        WriteOne(writer, 2);
+        cancel_done_server_barrier.Wait();
+        return grpc::Status::OK;
+      });
+
+  HandlerResult result;
+  btproto::MutateRowsRequest request;
+  auto context = google::cloud::internal::make_unique<grpc::ClientContext>();
+  auto op = cq_.MakeStreamingReadRpc(
+      [this](grpc::ClientContext* context,
+             btproto::MutateRowsRequest const& request,
+             grpc::CompletionQueue* cq) {
+        return stub_->PrepareAsyncMutateRows(context, request, cq);
+      },
+      request, std::move(context),
+      [&result](btproto::MutateRowsResponse r) {
+        result.reads.emplace_back(std::move(r));
+        return make_ready_future(true);
+      },
+      [&result](Status s) {
+        result.status = std::move(s);
+        result.done.Lift();
+      });
+
+  server_started_barrier.Wait();
+  op->Cancel();
+
+  // The server remains blocked until the stream finishes, therefore, the only
+  // way this actually unblocks is if the Cancel() succeeds.
+  result.done.Wait();
+  // There is no guarantee on how many messages will be received before the
```

Comment:
Perhaps check that it is 0 <= n <= 3?

Reply:
`0 <= n` is trivial because `n` is unsigned.  I added the `n <= 3` check though, good idea.